File: 02-observer/weather-node/displays/forecast.ts
import Subject from "../weatherdata/subject";
import DisplayElement from "./display";
import Observer from "./observer";

export default class ForecastDisplay implements Observer, DisplayElement {
  private currentPressure: number;
  private lastPressure: number;
  private weatherData: Subject;

  constructor(weatherData: Subject) {
    this.weatherData = weatherData;
    this.currentPressure = 29.92;
    this.lastPressure = 0;
    this.weatherData.registerObservers(this);
  }

  public update(temp: number, humidity: number, pressure: number): void {
    this.lastPressure = this.currentPressure;
    this.currentPressure = pressure;
    this.display();
  }

  public display(): void {
    let msg = "Forecast: ";
    if (this.currentPressure > this.lastPressure) {
      msg += "Improving weather on the way!";
    } else if (this.currentPressure == this.lastPressure) {
      msg += "More of the same";
    } else {
      msg += "Watch out for cooler, rainy weather";
    }

    console.log(msg);
  }
}

File: 02-observer/weather-go/display/forecast.go
package display

import (
	"fmt"

	"github.com/k-nox/design-patterns/02-observer/weather-go/observer"
)

type Forecast struct {
	currentPressure float32
	lastPressure float32
	weatherData observer.Subject
}

func (f *Forecast) Update(temp float32, humidity float32, pressure float32) {
	f.lastPressure = f.currentPressure
	f.currentPressure = pressure
	f.Display()
}

func (f *Forecast) Display() {
	fmt.Print("Forecast: ")
	if f.currentPressure > f.lastPressure {
		fmt.Println("Improving weather on the way!")
	} else if f.currentPressure == f.lastPressure {
		fmt.Println("More of the same")
	} else {
		fmt.Println("Watch out for cooler, rainy weather")
	}
}

func NewForecast(wd observer.Subject) *Forecast {
	f := &Forecast{
		currentPressure: 29.92,
		weatherData: wd,
	}

	wd.RegisterObserver(f)
	return f
}